### PR TITLE
Files ready for PR

### PR DESCRIPTION
**Pull request by AustinJRoberts**
Take 2 on this PR

### CODE REVIEW

**Comment by mcfarljm** [comment]
Suggest removing `RNA_plot.png` from version control, since it is created as an output of the code.  Currently, running the code or tests will modify the file and cause git to see it as modified.

File: RNA_folding.py [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+from itertools import product, combinations
+import numpy as np
+import networkx as nx
+import matplotlib.pyplot as plt
+import dimod
+from dwave.system import LeapHybridCQMSampler
+
+"""
+This file contains functions called  to help in the building and processing of models related to RNA folding.
+The models are built and solved in a different file.
```
Comment (mcfarljm) [inline RNA_folding.py]: Is this correct?  It looks like this file is self-contained.

File: RNA_folding.py [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+    Returns: Numpy matrix of 0's and 1's, where 1 represents a possible bonding pair.
+    """
+
+    # Requires text file of rna data written in same format as examples.
+    with open(file_name) as f:
+        rna = "".join(("".join(line.split()[1:]) for line in f.readlines())).lower()
+
+    # Create a dictionary of all indices where each nucleotide occurs.
+    # Mote: t vs u is mainly preference of database. While u is in rna, t is in dna.
+    index_dict = {'a': [], 'c': [], 'g': [], 't': [], 'u': []}
```
Comment (mcfarljm) [inline RNA_folding.py]: You might consider using `defaultdict(list)` here (from the `collections` package):
```suggestion
    index_dict = defaultdict(list)
```

File: RNA_folding.py [hunk 3]
```diff
@@ -0,0 +1,319 @@
...
+    return bonded_stems
+
+
+if __name__ == "__main__":
+    file = 'RNA_text_files/TMGMV_UPD-PK1.txt'
+    # file = 'RNA_text_files/NGF-L6.txt'
+    # file = 'RNA_text_files/STMV_UPD2-PK1.txt'
+    # file = 'RNA_text_files/NC_008516.txt'
+    # file = 'RNA_text_files/hiv.txt'
+    # file = 'RNA_text_files/simple.txt'
```
Comment (mcfarljm) [inline RNA_folding.py]: The `click` package is really useful for easily turning this into a command line option.  For example, see https://github.com/dwave-examples/crop-rotation

File: README.md [hunk 4]
```diff
@@ -1 +1,113 @@
...
+linearly weighted by the negative square of their length, k.
+
+2. Potential pseudoknots correspond to quadratic terms 
+with weight equal to the product of the two lengths 
+times a positive parameter c.
+
+3. Overlapping stems are disallowed, which is enforced by a constraint.
+
+
+![objective](https://latex.codecogs.com/svg.latex?\min_{\bf&space;x}&space;\&space;\sum_i&space;-k_ix_k&space;\&space;&plus;&space;\&space;c\sum_{i,j}&space;k_ik_jx_ix_j)
```
Comment (mcfarljm) [inline README.md]: Should this show `k_i^2` in the linear term?

File: RNA_folding.py [hunk 5]
```diff
@@ -0,0 +1,319 @@
...
+        # Check maximal stems first.
+        if check_overlap(stem1, stem2):
+            # If maximal stems overlap, compare list of smaller stems.
+            for stem_pair in product(stem_dict[stem1], stem_dict[stem2]):
+                if check_overlap(stem_pair[0], stem_pair[1]):
+                    cqm.add_constraint(dimod.quicksum([dimod.Binary(stem) for stem in stem_pair]) <= 1)
+
+    # Add constraint disallowing other overlapping stems.
+    # for stem_pairs in find_overlaps(stem_dict):
+    #     cqm.add_constraint(dimod.quicksum([dimod.Binary(stem) for stem in stem_pairs]) <= 1)
```
Comment (mcfarljm) [inline RNA_folding.py]: Why is this commented out?

File: RNA_folding.py [hunk 6]
```diff
@@ -0,0 +1,319 @@
...
+    for sample in sample_set.data():
+        if sample.is_feasible:
+            solution = sample
+            infeasible = False
+            break
+
+    # Check for feasible solution.
+    if infeasible:
+        print('\nWarning! All solutions infeasible. You may need to try again.')
+        return None
```
Comment (mcfarljm) [inline RNA_folding.py]: With latest dimod, this can be replaced with:
```suggestion
    feasible_samples = sample_set.filter(lambda s: s.is_feasible)
    if not feasible_samples:
        print('\nWarning! All solutions infeasible. You may need to try again.')
        return None
    best = feasible.first
```

Then use `best.energy` and `best.sample` in the later code.

File: RNA_folding.py [hunk 7]
```diff
@@ -0,0 +1,319 @@
...
+def make_stem_dict(bond_matrix, min_stem=3, min_loop=2):
+    """ Takes a matrix of potential hydrogen binding pairs and returns a dictionary of possible stems.
+
+    Each key is a maximal stem (under inclusion) whose value pair is a list of stems weakly contained in the key.
+    Recording stems in this manner allows for faster computations.
+
+    Args:
+        bond_matrix: Numpy matrix of 0's and 1's, where 1 represents a possible bonding pair.
+        min_stem: Integer minimum number of nucleotides between two bonding nucleotides.
+        min_loop: Integer minimum number of nucleotides between two bonding nucleotides.
```
Comment (mcfarljm) [inline RNA_folding.py]: Should one of these have a different description?

File: RNA_folding.py [hunk 8]
```diff
@@ -0,0 +1,319 @@
...
+    n = bond_matrix.shape[0]
+
+    # Iterate through matrix looking for possible stems.
+    for i in range(n - (2 * min_stem + min_loop)):
+        for j in range(i + 2 * min_stem + min_loop - 1, n):
+            if bond_matrix[i, j]:
+                k = 1
+                while bond_matrix[i + k, j - k]:
+                    bond_matrix[i + k, j - k] = False
+                    k += 1
```
Comment (mcfarljm) [inline RNA_folding.py]: I guess the content of `bond_matrix` is such that you will never run into a case where you try to access an element with `j - k < 0`?  You might consider reworking that to enforce it logically using a `for` loop so that avoiding the access error is not dependent on the array content:
```suggestion
                for k in range(1, j + 1):
                    if bond_matrix[i + k, j - k]:
                        bond_matrix[i + k, j - k] = False
                    else:
                        break
```

Comment (mcfarljm) [reply RNA_folding.py]: So now I'm seeing that `k` is used after the loop also.  That's still valid with this change, but arguably less readable to use `k` outside the `for` loop.  Up to you what to do here, this just jumped out at me as a situation where it's not obvious that you're preventing an array access on a negative index on line 89.

File: RNA_folding.py (hunk 1 above)
Comment (AustinJRoberts) [reply RNA_folding.py]: Correct. That was a left over from when I had a proper split between BQM and CQM. I'll delete it.

File: RNA_folding.py (hunk 7 above)
Comment (AustinJRoberts) [reply RNA_folding.py]: Yep. Thanks.

File: README.md (hunk 4 above)
Comment (mcfarljm) [reply README.md]: ~Also, is every pair of stems a potential pseudoknot?  The objective function shows coefficients for every pair, but it looks like the code in `pseudoknot_terms` does some filtering.~

Edit: This is probably OK, but I think the disconnect is in the fact that Fox (2021) define the objective as a square of sum of stem lengths, which produces interactions -- see my separate comment in the code where you implement the linear coefficients.

File: RNA_folding.py [hunk 9]
```diff
@@ -0,0 +1,319 @@
...
+
+    Args:
+        stem_dict: Dictionary with maximal stems as keys and list of weakly contained substems as values.
+        min_stem: Integer minimum number of nucleotides between two bonding nucleotides.
+
+    Returns: Constrained Binary Model.
+    """
+
+    # Create linear coefficients the prioritize inclusion of long stems.
+    linear_coeffs = {stem: -1 * (stem[1] - stem[0] + 1) ** 2 for sublist in stem_dict.values() for stem in sublist}
```
Comment (mcfarljm) [inline RNA_folding.py]: After looking at the Fox (2021) paper a little bit more closely, there seems to be a disconnect, at least in my understanding.  Their Eq. (3) is based on squaring the sum, not the sum of squares, which is what you're showing here (or at least the way the comment explains it).  Squaring the sum does produce the linear terms you show here, but it also produces interaction terms, as shown in Eq. (4).

In other words, the way I read their description, scoring consecutive base pairs (prioritizing long stems) includes both linear and quadratic terms due to the fact that they are measuring it using a square of the sum (and side note, not clear to me why they use a square instead of just a sum, but that's a separate question).

Comment (mcfarljm) [reply RNA_folding.py]: I think this observation is also born out in the energy values of the solutions.  I made a new test case based on the example from Fox (2021, Figure 1).  The solution found from your code has two length-3 stems.  No pseudoknot terms are present.  Based on Fox (2021), Eq. (3), I would expect the energy of this solution to be `-(3 + 3)^2 = -36`.  However, the code returns `-18`, which I believe is because you've implemented the objective as a sum of squares, not square of sum.

FYI, here's the input I was using.  Might also be worth including this in the code, since it can be referenced against the excellent diagrams that Fox (2021) provide in Figure 1.

```
1 ggaagcaaacaucccugu
```

File: README.md [hunk 10]
```diff
@@ -1 +1,113 @@
...
+Possible stems (each corresponding to a decision variable) 
+are stored in a dictionary structure to speed processing.
+
+## Code Specifics
+
+By default, the minimum stem length is set to 3. 
+A stem of length 5 thus contains
+two stems of length 4 and three stems of length 3 under inclusion.
+Our stem dictionary records each maximal stems (under inclusion) as keys
+and a list of all stems it weakly contains as the value pair.
```
Comment (mcfarljm) [inline README.md]: Is "value pair" the right term here?  What is the pair?  I think this came from key/value pair, but in this usage "pair" seems to describe the value itself, which is not a pair but a list.  Think it would be more clear as:
```suggestion
Our stem dictionary records the maximal stems (under inclusion) as keys,
where each key maps to a list of the associated stems weakly contained within the maximal stem.
```

File: RNA_folding.py [hunk 11]
```diff
@@ -0,0 +1,319 @@
...
+            if abs(bond[0] - bond[1]) > min_loop:
+                bond_matrix[min(bond), max(bond)] = 1
+
+    return bond_matrix
+
+
+def make_stem_dict(bond_matrix, min_stem=3, min_loop=2):
+    """ Takes a matrix of potential hydrogen binding pairs and returns a dictionary of possible stems.
+
+    Each key is a maximal stem (under inclusion) whose value pair is a list of stems weakly contained in the key.
```
Comment (mcfarljm) [inline RNA_folding.py]: See my comment in the README about usage of "value pair", which seems confusing here.

File: RNA_folding.py [hunk 12]
```diff
@@ -0,0 +1,319 @@
...
+    labels = {i: rna[i].upper() for i in range(len(rna))}
+    nx.draw_networkx_labels(G, pos, labels, font_size=10, font_color="whitesmoke")
+
+    nx.draw_networkx_edges(G, pos, edgelist=rna_edges, width=3.0, alpha=0.5)
+    nx.draw_networkx_edges(G, pos, edgelist=stem_edges, width=4.5, alpha=0.8, edge_color='tab:pink')
+
+    plt.savefig(fig_name + '.png')
+
+
+def build_cqm(stem_dict, min_stem=3):
```
Comment (mcfarljm) [inline RNA_folding.py]: Looks like `min_stem` is not used in the function

File: RNA_folding.py [hunk 13]
```diff
@@ -0,0 +1,319 @@
...
+
+    Returns: Dictionary with all possible pseudoknots as keys and appropriate penalty as as value pair.
+    """
+
+    pseudos = {}
+    # Look within all pairs of maximal stems for possible psuedoknots.
+    for stem1, stem2 in product(stem_dict.keys(), stem_dict.keys()):
+        # Using product instead of combinations allows for short asymmetric checks.
+        if stem1[0] + 2 * min_stem < stem2[1] and stem1[2] + 2 * min_stem < stem2[3]:
+            pseudos.update({(substem1, substem2): c * (substem1[1] - substem1[0]) * (substem2[1] - substem2[0])
```
Comment (mcfarljm) [inline RNA_folding.py]: Do you need to add 1 when computing the stem lengths, like you do in other places?
```suggestion
            pseudos.update({(substem1, substem2): c * (substem1[1] - substem1[0] + 1) * (substem2[1] - substem2[0] + 1)
```

File: RNA_folding.py (hunk 2 above)
Comment (AustinJRoberts) [reply RNA_folding.py]: Good call.

File: RNA_folding.py (hunk 3 above)
Comment (AustinJRoberts) [reply RNA_folding.py]: I'll check it out. I wasn't looking forward to using argpars anyway.

File: README.md (hunk 4 above)
Comment (AustinJRoberts) [reply README.md]: This may be a longer conversation.

File: RNA_folding.py (hunk 5 above)
Comment (AustinJRoberts) [reply RNA_folding.py]: Legacy code from some comparisons I was running. I'll delete it.

File: RNA_folding.py (hunk 6 above)
Comment (AustinJRoberts) [reply RNA_folding.py]: Nice.

File: RNA_folding.py (hunk 8 above)
Comment (AustinJRoberts) [reply RNA_folding.py]: I think I would prefer to add a comment about the strictly upper triangular nature of the matrix. For loops are slow and in this case less readable.

File: RNA_folding.py (hunk 9 above)
Comment (AustinJRoberts) [reply RNA_folding.py]: This may be a longer conversation. Their paper is not internally consistent on this point, so I took some liberties. Will follow up outside of PR.

File: RNA_folding.py (hunk 13 above)
Comment (AustinJRoberts) [reply RNA_folding.py]: Yeah... I suppose you're right. Ugly code, cleaner concept.

File: RNA_folding.py (hunk 12 above)
Comment (AustinJRoberts) [reply RNA_folding.py]: Yep. I'll take it out ... but I may need to add it back in if I expose min_stem to the user.

File: RNA_folding.py (hunk 9 above)
Comment (mcfarljm) [reply RNA_folding.py]: Agree, let's discuss.  One thing that definitely looked off to me in the paper was the transition from Eq. (6) to (8).

File: README.md [hunk 14]
```diff
@@ -1 +1,113 @@
...
+
+In biology and chemistry,
+the properties of a molecule are not solely determined by a set of atoms
+but also by the shape of the molecule. 
+In genetics, the shape of an RNA molecule is largely determined by how it bends back on itself. 
+The sequence of A’s, U’s, G’s, and C’s that make up RNA has 
+certain pairs are drawn together to form hydrogen bonds.
+several bonds in a row is called a stem,
+and a stem provides sufficient force to keep the molecule folded together.
+This demo program takes a sequence of RNA and applies a quadratic model to find the optimal stem configuration.
```
Comment (mcfarljm) [inline README.md]: Can you be more specific about optimal here?  What is the problem that's really being solved?  Are we looking for the lowest energy configuration?

File: README.md (hunk 4 above)
Comment (mcfarljm) [reply README.md]: Following up from yesterday's conversation, it sounds like there are still two updates that need to be made to this equation:

1. Linear term should have `k_i^2`
2. Need to express that quadratic terms apply only to stem pairs that are associated with pseudoknots.  E.g., either use a set notation for the sum or use a delta term.

File: README.md [hunk 15]
```diff
@@ -1 +1,113 @@
...
+
+## Code Specifics
+
+By default, the minimum stem length is set to 3. 
+A stem of length 5 thus contains
+two stems of length 4 and three stems of length 3 under inclusion.
+The stem dictionary records the maximal stems (under inclusion) as keys,
+where each key maps to a list of the associated stems weakly contained within the maximal stem.
+
+No two stems contained in the same maximal key can both be in an optimal solution,
```
Comment (mcfarljm) [inline README.md:135]: This seems right, but just wanted to double-check how sure you are about this.  Suppose that we had a fairly large maximal stem.  Is it possible that it would be optimal to choose multiple substems within it, due to pseudknots or the `min_loop` requirement?

File: RNA_folding.py [hunk 16]
```diff
@@ -0,0 +1,319 @@
...
+        min_loop: Integer minimum number of nucleotides between two bonding nucleotides.
+
+    Returns: Dictionary of all possible stems with maximal stems as keys.
+    """
+
+    stem_dict = {}
+    n = bond_matrix.shape[0]
+
+    # Iterate through matrix looking for possible stems.
+    for i in range(n - (2 * min_stem + min_loop)):
```
Comment (mcfarljm) [inline RNA_folding.py]: Should this have a `+1`?  For example, suppose I have `n=10` nuceotides, `min_stem=5`, and `min_loop=0`.  This will produce an empty `for` loop, but is the length-5 stem not valid?
```suggestion
    for i in range(n - (2 * min_stem + min_loop) + 1):
```

Comment (mcfarljm) [reply RNA_folding.py]: Here's an example input:
```
1 aaaaa ttttt
```

This produces a solution with `min_stem=4` and `min_loop=0`, but with `min_stem=5` and `min_loop=0` it creates an exception because there are no variables (side note, could consider gracefully handling the case where there are no variables).

Comment (mcfarljm) [reply RNA_folding.py]: Hmm, maybe `min_loop=0` doesn't actually make sense because you can't have a hydrogen bond for two nucleotides that are already in sequence.

However, it still looks like there is an range issue here.  The above example input produces a solution with `min_stem=4` and `min_loop=1` that, if I understand correctly, has a loop of 2.  However, running with `min_stem=4` and `min_loop=2` produces an exception.

File: RNA_folding.py [hunk 17]
```diff
@@ -0,0 +1,319 @@
...
+@click.command(help='Solve an instance of the RNA folding problem using '
+                    'LeapHybridCQMSampler.')
+@click.option('--path', type=click.Path(), default=DEFAULT_PATH,
+              help=f'Path to problem file.  Default is {DEFAULT_PATH!r}')
+@click.option('--verbose', is_flag=True, default=True)
+def main(path, verbose):
+    if verbose:
+        print('\nPreprocessing data from:', path)
+
+    stem_dict = make_stem_dict(text_to_matrix(path))
```
Comment (mcfarljm) [inline RNA_folding.py]: This would be a good place to trap the case where there are no stems.  Believe you could just do `if not stem_dict` and then print an error and return.  Otherwise it leads to an exception getting raised in the solver.

File: README.md (hunk 15 above)
Comment (AustinJRoberts) [reply README.md:135]: I'm sure. The only thing worth considering is if we split a stem in order to fit another stem in between the two parts, which will always be less good (unless we decide to put large negative coefficient on pseudoknots for no particular reason ... so maybe I should explicitly not allow that)

Comment (mcfarljm) [reply README.md:135]: 👍

File: README.md [hunk 18]
```diff
@@ -1 +1,154 @@
...
+
+<p align = "center">
+
+![Figure 2](readme_imgs/pseudoknot2.png)
+<p align = "center">
+Fig.2 - A pseudoknot formed by a stem of length 3 and a stem of length 5.
+</p>
+
+This demo is loosely based on the work of the first paper cited in the References section,
+which is in turn inspired by the second paper in the Reference section.
```
Comment (mcfarljm) [inline README.md]: Probably a good point to go ahead and switch over to actual citations.  You can check the other examples, but I think we've typically used numbered citations.

File: RNA_folding.py [hunk 19]
```diff
@@ -0,0 +1,329 @@
...
+# Create command line functionality.
+DEFAULT_PATH = join(dirname(__file__), 'RNA_text_files', 'TMGMV_UPD-PK1.txt')
+
+
+@click.command(help='Solve an instance of the RNA folding problem using '
+                    'LeapHybridCQMSampler.')
+@click.option('--path', type=click.Path(), default=DEFAULT_PATH,
+              help=f'Path to problem file.  Default is {DEFAULT_PATH!r}')
+@click.option('--verbose', is_flag=True, default=True)
+@click.option('--min_stem', type=click.INT, default=50)
```
Comment (mcfarljm) [inline RNA_folding.py]: Did you mean to set this to 50?  The default problem does not seem to work with this setting.

File: RNA_folding.py [hunk 20]
```diff
@@ -0,0 +1,329 @@
...
+@click.option('--min_loop', type=click.INT, default=2)
```
Comment (mcfarljm) [inline RNA_folding.py]: It's minor, but conventionally we use dashes instead of underscores here.  If nothing else, it's easier to type.  (They'll get converted to underscores for the actual variable names internally.)

File: RNA_folding.py [hunk 21]
```diff
@@ -0,0 +1,329 @@
...
+        verbose: Boolean indicating if function should print additional information.
+
+    Returns: List of stems included in optimal solution, encoded as 4-tuples.
+    """
+
+    # Filter for feasibility
+    feasible_samples = sample_set.filter(lambda s: s.is_feasible)
+    # Check that feasible example exists
+    if not feasible_samples:
+        print('\033[93m' + '\nWarning! All solutions infeasible. You may need to try again.')
```
Comment (mcfarljm) [inline RNA_folding.py]: You've got some interesting characters here.  Not sure what they are or if it was intentional, but they don't seem to display anything on my terminal.

File: RNA_folding.py [hunk 22]
```diff
@@ -0,0 +1,329 @@
...
+        return None
```
Comment (mcfarljm) [inline RNA_folding.py]: It looks like this returning `None` here leads to a later exception down the line when calling `make_plots`.  You could raise an exception here, call `sys.exit()`, or handle the `None` case in `main`.

Btw, an easy way to test the behavior is to just add `or True` to the if statement.

File: README.md (hunk 14 above)
Comment (AustinJRoberts) [reply README.md]: Added a mention of energy.

File: RNA_folding.py (hunk 21 above)
Comment (AustinJRoberts) [reply RNA_folding.py]: They set the color to 'Warning yellow'. I will get on a more elegant solution.

File: RNA_folding.py (hunk 22 above)
Comment (AustinJRoberts) [reply RNA_folding.py]: I always forget about sys.exit()

File: RNA_folding.py (hunk 16 above)
Comment (AustinJRoberts) [reply RNA_folding.py]: I believe it was at the intersection of a +1 on the range and some sanitization of inputs. Hopefully people are okay with using 'assert' for that purpose.

File: RNA_folding.py [hunk 23]
```diff
@@ -0,0 +1,355 @@
...
+    """
+    # Sanitize inputs
+    assert (min_stem >= 1), 'min_stem must be greater or equal to 1.'
+    assert (min_loop >= 0), 'min_loop must be non-negative.'
+    assert (c >= 0), 'c must be non-negative'
+
+    if verbose:
+        print('\nPreprocessing data from:', path)
+
+    stem_dict = make_stem_dict(text_to_matrix(path), min_stem, min_loop)
```
Comment (mcfarljm) [inline RNA_folding.py]: Do you need to pass the value of `min_loop` into `text_to_matrix`?
```suggestion
    stem_dict = make_stem_dict(text_to_matrix(path, min_loop), min_stem, min_loop)
```

File: RNA_folding.py [hunk 24]
```diff
@@ -0,0 +1,355 @@
...
+        min_stem: Integer smallest number of consecutive bonds to be considered a stem.
+        min_loop: Integer minimum number of nucleotides separating two sides of a stem.
+        c: Float multiplier for the coefficient of the quadratic terms for pseudoknots.
+
+    Returns:
+    """
+    # Sanitize inputs
+    assert (min_stem >= 1), 'min_stem must be greater or equal to 1.'
+    assert (min_loop >= 0), 'min_loop must be non-negative.'
+    assert (c >= 0), 'c must be non-negative'
```
Comment (mcfarljm) [inline RNA_folding.py]: You can actually handle all of these cases directly with `click`: https://click.palletsprojects.com/en/8.0.x/options/#range-options.  For example:
```python
@click.option('--min-stem', type=click.IntRange(1), default=3,
              help='Minimum length for a stem to be considered.')
```

File: RNA_folding.py [hunk 25]
```diff
@@ -0,0 +1,349 @@
...
+
+    sampler = LeapHybridCQMSampler()
+
+    if verbose:
+        print('Finding Solution...')
+
+    sample_set = sampler.sample_cqm(cqm)
+
+    if verbose:
+        print('Processing solution...')
```
Comment (mcfarljm) [inline RNA_folding.py:344]: The output here is a little bit misleading.  The value returned by `sampler.sample_cqm` is a lazy object that doesn't actually evaluate until you access it.  This print statement implies that "Finding Solution" has completed, but that's not really the case.  So if you're watching the screen output in realtime, you get the impression that "Finding solution" is very fast and most time is spent in "Processing solution", but that's not true.

You haven't really obtained a solution until you force the lazy object to be evaluated, which doesn't happen until line 259 that filters the sampleset.  If you put a print statement immediately after that, you will see that what really happens is that most of the program time is spent in waiting for the solution to return (i.e., "finding the solution").

Comment (AustinJRoberts) [reply RNA_folding.py:344]: Would it make sense to add sample_set.resolve() before that print statement?

Comment (mcfarljm) [reply RNA_folding.py:344]: Yup, that would probably be the easiest thing to do here.

File: README.md [hunk 26]
```diff
@@ -1 +1,154 @@
...
+but also by the shape of the molecule. 
+In genetics, the shape of an RNA molecule is largely determined by how it bends back on itself. 
+The sequence of A’s, U’s, G’s, and C’s that make up RNA has 
+certain pairs are drawn together to form hydrogen bonds.
+A sequence of several bonds in a row is called a stem,
+and a stem provides sufficient force to keep the molecule folded together.
+In nature, an RNA molecule will form some stems while avoiding others
+in a manner that  minimizes the free energy of the system. 
+
+This demo program takes am RNA sequence and applies a quadratic model in persuit of the optimal stem configuration.
```
Comment (mcfarljm) [inline README.md]:
```suggestion
This demo program takes am RNA sequence and applies a quadratic model in pursuit of the optimal stem configuration.
```

File: README.md [hunk 27]
```diff
@@ -1 +1,154 @@
...
+The stem expressed by the pink lines can be denoted with the tuple (2, 5, 14, 17),
+where 2 is the index of the first G,
+4 is the index of the U,
+14 is the index of the A,
+and 17 is the index of the last C.
+This 4-tuple is mapped to a single variable.
+However, the smaller stems (2, 4, 15, 17) and (3, 5, 15, 16) also need to be considered,
+even though the optimal solution will not include them in this case.
+
+Second, the intertwining phenomenon known as psuedoknots are less energetically favorable.
```
Comment (mcfarljm) [inline README.md]:
```suggestion
Second, the intertwining phenomenon known as pseudoknots are less energetically favorable.
```

File: README.md [hunk 28]
```diff
@@ -1 +1,154 @@
...
+
+## Usage
+
+To run the demo through a command line interface, type:
+
+```bash
+python RNA_folding.py
+```
+
+The demo prints the optimal stem configuration along with other relavent data.
```
Comment (mcfarljm) [inline README.md]:
```suggestion
The demo prints the optimal stem configuration along with other relevant data.
```

File: README.md [hunk 29]
```diff
@@ -1 +1,154 @@
...
+It then saves a plot of the sequence and its bonds as `RNA_plot.png`.
+
+### Optional parameters
+Several optional parameters are accepted:
+
+- `--path`: specifies path the input text file with RNA sequence information. 
+- `--verbose`: if set to default value of 'True',
+the program prints additional information about the model. 
+- `--min-stem`: minimum length necessary for a stem to be considered.
+- `--min-loop`: minimum number of neucleotides that must be present
```
Comment (mcfarljm) [inline README.md]:
```suggestion
- `--min-loop`: minimum number of nucleotides that must be present
```

File: README.md [hunk 30]
```diff
@@ -1 +1,154 @@
...
+3. Overlapping stems are disallowed, which is enforced by a constraint.
+
+![objective](readme_imgs/objective.png) 
+
+Subject to ![constraint](readme_imgs/constraint.png) if stems *i* and *j* overlap.
+
+Here, each *x<sub>i</sub>* is a binary variable indicating the inclusion/exclusion of the *i<sup>th</sup>* stem.
+Each constant *k<sub>i</sub>* is the length of said stem.
+The indexing set *S* is the set of all pairs of stems that forma a pseudoknot.
+Finally, *c* is a tunable parameter adjusting the impact of pseudonknots.
```
Comment (mcfarljm) [inline README.md]:
```suggestion
The indexing set *S* is the set of all pairs of stems that form a pseudoknot.
Finally, *c* is a tunable parameter adjusting the impact of pseudoknots.
```

File: README.md [hunk 31]
```diff
@@ -1 +1,154 @@
...
+Predicting the existence of stems is important to predicting the properties of the RNA molecule.
+However, prediction is made more complicated by two important factors. 
+
+First, stems are not allowed to overlap. 
+A simple case of overlapping can be illustrated by Figure 1.
+The stem expressed by the pink lines can be denoted with the tuple (2, 5, 14, 17),
+where 2 is the index of the first G,
+4 is the index of the U,
+14 is the index of the A,
+and 17 is the index of the last C.
```
Comment (mcfarljm) [inline README.md]: I'm having trouble reconciling this with the figure:

- Should the text say "5 is the index of the U"?
- Isn't ~11~ 12 the index of the A?
- Is 15 the index of the last C in the stem?  If so, "last C" might be confusing in the text because it's not the last C in the full sequence.

File: README.md [hunk 32]
```diff
@@ -1 +1,154 @@
...
+
+Second, the intertwining phenomenon known as pseudoknots are less energetically favorable.
+In Figure 2, we see an example of such a pseudoknot, 
+where one side of a stem occurs in between the two sides of a different stem.
+The use of a quadratic objective allows us to make pseudoknots less likely to occur in optimal solutions,
+increasing overall accuracy.
+
+<p align = "center">
+
+![Figure 2](readme_imgs/pseudoknot2.png)
```
Comment (mcfarljm) [inline README.md:45]: Looks like this image file is missing in the repo

File: .gitignore [hunk 33]
```diff
@@ -0,0 +1,2 @@
+RNA_BQM.py
+readme_imgs/pseudoknot2.png
```
Comment (mcfarljm) [inline .gitignore]: Is this supposed to be included in the repo instead of ignored?

File: README.md (hunk 31 above)
Comment (AustinJRoberts) [reply README.md]: Good catch. I changed the file when I redid the image parameters and forgot to update the readme.

File: README.md (hunk 32 above)
Comment (AustinJRoberts) [reply README.md:45]: hmmm.... I'll make sure that gets pushed in.

File: .gitignore (hunk 33 above)
Comment (AustinJRoberts) [reply .gitignore]: Yeah, don't know how I managed to do that.

File: tests/test_RNA_folding.py [hunk 34]
```diff
@@ -0,0 +1,59 @@
...
+        subprocess.check_output([sys.executable, demo_file])
+
+
+class TestRNA_folding(unittest.TestCase):
+    def test_num_vars(self):
+        """Test CQM characteristics for demo instance"""
+
+        file = os.path.join(project_dir, 'RNA_text_files/TMGMV_UPD-PK1.txt')
+
+        stem_dict = RNA_folding.make_stem_dict(RNA_folding.text_to_matrix(file))
```
Comment (mcfarljm) [inline tests/test_RNA_folding.py]: Missing `min_loop` argument.  Always a good idea to run your unit tests after making changes.

File: tests/test_RNA_folding.py [hunk 35]
```diff
@@ -0,0 +1,59 @@
...
+
+        self.assertEqual(len(cqm.variables), 17)
+        self.assertEqual(len(cqm.constraints), 36)
+
+    def test_small_case(self):
+        """Test solution quality of small case."""
+
+        file = os.path.join(project_dir, 'RNA_text_files/simple_pseudo.txt')
+
+        stem_dict = RNA_folding.make_stem_dict(RNA_folding.text_to_matrix(file))
```
Comment (mcfarljm) [inline tests/test_RNA_folding.py]: Missing `min_loop` argument.

**Review by mcfarljm — APPROVED** [review]
Great work.  Really neat example, and not an easy one, but you've done a nice job with both the implementation and the documentation.
